File: core/src/main/java/io/github/shaksternano/borgar/core/media/writerold/Image4jIcoWriter.java
package io.github.shaksternano.borgar.core.media.writerold;

import com.sksamuel.scrimage.ImmutableImage;
import io.github.shaksternano.borgar.core.media.ImageFrameOld;
import net.ifok.image.image4j.codec.ico.ICOEncoder;

import java.io.File;
import java.io.IOException;

public class Image4jIcoWriter implements NoAudioWriter {

    private final File output;
    private boolean written = false;

    public Image4jIcoWriter(File output) {
        this.output = output;
    }

    @Override
    public void writeImageFrame(ImageFrameOld frame) throws IOException {
        if (!written) {
            written = true;
            var image = ImmutableImage.wrapAwt(frame.getContent())
                .bound(256, 256)
                .awt();
            ICOEncoder.write(image, output);
        }
    }

    @Override
    public boolean isStatic() {
        return true;
    }

    @Override
    public void close() {
    }
}

File: core/src/main/kotlin/io/github/shaksternano/borgar/core/media/writer/Image4jIcoWriter.kt
package io.github.shaksternano.borgar.core.media.writer

import io.github.shaksternano.borgar.core.media.ImageFrame
import io.github.shaksternano.borgar.core.media.MediaWriterFactory
import io.github.shaksternano.borgar.core.media.bound
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import net.ifok.image.image4j.codec.ico.ICOEncoder
import java.nio.file.Path
import kotlin.time.Duration

private const val MAX_DIMENSION = 256

class Image4jIcoWriter(
    private val output: Path,
) : NoAudioWriter() {

    override val isStatic: Boolean = true
    private var written: Boolean = false

    override suspend fun writeImageFrame(frame: ImageFrame) {
        if (written) return
        written = true
        val image = frame.content.bound(MAX_DIMENSION)
        withContext(Dispatchers.IO) {
            ICOEncoder.write(image, output.toFile())
        }
    }

    override suspend fun close() = Unit

    object Factory : MediaWriterFactory {
        override val supportedFormats: Set<String> = setOf(
            "ico",
        )

        override suspend fun create(
            output: Path,
            outputFormat: String,
            loopCount: Int,
            audioChannels: Int,
            audioSampleRate: Int,
            audioBitrate: Int,
            maxFileSize: Long,
            maxDuration: Duration
        ): MediaWriter = Image4jIcoWriter(output)
    }
}
